spark3/src/main/java/org/apache/iceberg/actions/Spark3CreateAction.java

```diff
@@ -0,0 +1,172 @@
+  protected static final String ICEBERG_METADATA_FOLDER = "metadata";
+
+  private final SparkSession spark;
+
+  // Source Fields
+  private final V1Table sourceTable;
+  private final CatalogTable sourceCatalogTable;
+  private final String sourceTableLocation;
+  private final CatalogPlugin sourceCatalog;
+  private final Identifier sourceTableIdent;
+
+  // Destination Fields
+  private final StagingTableCatalog destCatalog;
+  private final Identifier destTableIdent;
+
+  // Optional Parameters for destination
+  private Map<String, String> additionalProperties = Maps.newHashMap();
+
+  Spark3CreateAction(SparkSession spark, CatalogPlugin sourceCatalog, Identifier sourceTableIdent,
+                       CatalogPlugin destCatalog, Identifier destTableIdent) {
+
+    this.spark = spark;
+    this.sourceCatalog = checkSourceCatalog(sourceCatalog);
+    this.sourceTableIdent = sourceTableIdent;
+    this.destCatalog = checkDestinationCatalog(destCatalog);
+    this.destTableIdent = destTableIdent;
+
+    try {
+      this.sourceTable = (V1Table) ((TableCatalog) sourceCatalog).loadTable(sourceTableIdent);
+      this.sourceCatalogTable = sourceTable.v1Table();
+    } catch (NoSuchTableException e) {
+      throw new IllegalArgumentException(String.format("Cannot not find source table %s", sourceTableIdent), e);
+    } catch (ClassCastException e) {
+      throw new IllegalArgumentException(String.format("Cannot use non-v1 table %s as a source", sourceTableIdent), e);
+    }
+    validateSourceTable(sourceCatalogTable);
+
+    this.sourceTableLocation = CatalogUtils.URIToString(sourceCatalogTable.storage().locationUri().get());
+  }
+
+  @Override
+  public CreateAction withProperties(Map<String, String> properties) {
+    this.additionalProperties.putAll(properties);
+    return this;
+  }
+
+  @Override
+  public CreateAction withProperty(String key, String value) {
+    this.additionalProperties.put(key, value);
+    return this;
+  }
+
+  protected SparkSession spark() {
+    return spark;
+  }
+
+  protected String sourceTableLocation() {
+    return sourceTableLocation;
+  }
+
+  protected CatalogTable v1SourceTable() {
+    return sourceCatalogTable;
+  }
+
+  protected CatalogPlugin sourceCatalog() {
+    return sourceCatalog;
+  }
+
+  protected Identifier sourceTableName() {
+    return sourceTableIdent;
+  }
+
+  protected Transform[] sourcePartitionSpec() {
+    return sourceTable.partitioning();
+  }
+
+  protected StagingTableCatalog destCatalog() {
+    return destCatalog;
+  }
+
+  protected Identifier destTableName() {
+    return destTableIdent;
+  }
+
+  protected Map<String, String> additionalProperties() {
+    return additionalProperties;
+  }
+
+  private static void validateSourceTable(CatalogTable sourceTable) {
+    String sourceTableProvider = sourceTable.provider().get().toLowerCase(Locale.ROOT);
+
+    if (!ALLOWED_SOURCES.contains(sourceTableProvider)) {
```

Comment:
I think it will be more readable to use `Preconditions` here.

```
    Preconditions.checkArgument(ALLOWED_SOURCES.contains(sourceTableProvider),
        "Cannot create an Iceberg table from source provider: %s",
        sourceTableProvider);
```